### DISCUSSION TITLE
What's the general consensus for handling multiple kinds of errors for commands in Rust?

### DISCUSSION THREAD

**Discussion by bigbeno37** [tauri-apps/tauri#5008] [Q&A]
Hey everyone, I'm currently writing a Tauri app to handle certain interactions with the filesystem, and this of course means plenty of Results are coming my way. For instance, I'd like to read a file from the file system, and then use serde_json to parse it, resulting in potential io::Error and serde_json::Error errors.

```rust
struct SomeJSON {
    // ...
}

#[tauri::command]
fn get_json_file() -> Result<SomeJSON, ???> {
    let path = Path::new("./file.json");
    let contents = read_to_string(&path)?;
    serde_json::from_str(contents)
}
```

I noticed tauri::Error implements a lot of From traits for various packages, so my command function could return a Result<T, tauri::Error>, but I now have a problem if I use an unsupported Error, such as using the Zip crate, which produces its own ZipErrors.

```rust
struct SomeJSON {
    // ...
}

#[tauri::command]
fn get_json_file() -> Result<SomeJSON, tauri::Error> {
    let path = Path::new("./file.json");
    let file = File::open(&path)?;
    // oops...`?` couldn't convert the error to `tauri::Error`
    let archive = zip::ZipArchive::new(file)?;
    // go through archive and get json file...
}
```

I looked into the anyhow crate to see if it could help, but if my command function returned an anyhow::Error, the macro errors out, complaining that anyhow::Error doesn't implement serde::Serialize, which makes sense.

Is there a recommended way of handling multiple types of Errors? What have you implemented for error handling in your own Tauri apps?

**Comment by FabianLars** [comment]
We started exploring this topic in our WIP docs: <https://jonaskruckenberg.github.io/tauri-docs-wip/development/inter-process-communication.html#error-handling>.
I guess you could call that our recommendation for now 😅
